proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+ *
+ * For `silenceDeprecations`, we chose to make passing anything other than
+ * an active deprecation an error to prevent users from unintentionally
+ * over-silencing. Future deprecations are already effectively silenced, so
+ * users shouldn't be able to pass it here until it's actually active. If the
+ * deprecation a user was silencing is now obsolete, that means the breaking
+ * change has now been made, so silencing has no effect, and users who were
+ * silencing it may not be ready for the actual breaking change. There is no
+ * way to distinguish between different user-authored deprecations, so they
+ * should not be able to be silenced as a group.
+ *
+ * For `futureDeprecations`, we only emit a warning for non-future deprecations,
+ * since future deprecations all eventually become active. The warning will let
+ * users know to clean up their configuration, but there should be no behavior
+ * change.
+ */
+
+/* ## API */
+
+import {SourceSpan} from '../spec/js-api';
+
+declare module '../spec/js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    /**
+     * A set of deprecations to treat as fatal.
+     *
+     * If a version is provided, then all deprecations released in that
+     * version or earlier will be treated as fatal. If a string is provided, it
+     * will resolve to the deprecation with that ID, or error if none exists.
+     *
+     * This will throw an error if a future deprecation is included here, unless
+     * that future deprecation is also passed to `futureDeprecations`. This
+     * will emit a warning if an obsolete deprecation is included here. */
+    fatalDeprecations?: (Deprecation | string | Version)[];
+
+    /**
+     * A set of active deprecations to ignore.
+     *
+     * If a string is provided, it will resolve to the deprecation with that ID,
+     * or error if none exists.
+     *
+     * This will throw an error if a future, obsolete, or user-authored
+     * deprecation is included here.
+     */
+    silenceDeprecations?: (Deprecation | string)[];
+
+    /**
+     * A set of future deprecations to opt into early.
+     *
+     * If a string is provided, it will resolve to the deprecation with that ID,
+     * or error if none exists.
+     *
+     * This will emit a warning if a non-future deprecation is included here.
+     */
+    futureDeprecations?: (Deprecation | string)[];
+  }
+
+  interface Logger {
+    /**
+     * Update the third sub-bullet of bullet two to read:
+     *
+     * If this warning is caused by behavior that used to be allowed but will
+     * be disallowed in the future, set `options.deprecation` to `true` and
+     * set `options.deprecationType` to the relevant `Deprecation`. Otherwise,
+     * set `options.deprecation` to `false` and leave `options.deprecationType`
+     * undefined.
+     */
+    warn?(
+      message: string,
+      options: {
+        deprecation: boolean;
+        deprecationType?: Deprecation;
```

Comment:
I thought for the Dart API we decided that it was better to not allow user-authored warnings to have specific deprecation types, and to lump them all under "user deprecations"